### PR TITLE
initial implementation of weight feature for positioning of newslette…

### PR DESCRIPTION
**Pull request by svenow**
each field in contact section of form has an additional weight field. the weight can be set to values between 0 and 100. the default weight is 50.

a help text exists (when pressing the small blue question mark besides the weight input field) for a brief description about using weights.

what might be missing is setting the default weight-value for each field in setDefaultValues(), if weight has not been set yet. currently, this is not done and the weight value that is delivered via the api might be empty.

max and default weight values are defined as public static class variables, in order to access them in smarty-templates.

*systopia-reference: 24994*

### CODE REVIEW

File: CRM/Newsletter/Form/Profile.php [hunk 1]
```diff
@@ -188,6 +202,19 @@ public function buildQuickForm() {
...
+      $this->add(
+        'text',
+        'contact_field_' . $contact_field_name . '_weight',
+        E::ts('Field position'),
+        array('placeholder' => E::ts(
+	  'Default weight: %1',
+          array(
+            1 => static::$_form_field_weight_default
+          )
+        ))
```
Comment (jensschuppe) [inline CRM/Newsletter/Form/Profile.php:214]: Let's use short array syntax `[]` instead of `array()`.

File: CRM/Newsletter/Form/Profile.php [hunk 2]
```diff
@@ -44,6 +44,20 @@ class CRM_Newsletter_Form_Profile extends CRM_Core_Form {
...
+   * The default weight of form fields, for positioning within the form
+   */
+  public static $_form_field_weight_default = 50;
+  
+  /**
+   * @var string
+   *
+   * The max weight value for form fields
+   */
+  public static $_form_field_weight_max = 100;
```
Comment (jensschuppe) [inline CRM/Newsletter/Form/Profile.php:60]: Those should be class constants.

File: CRM/Newsletter/Form/Profile.php [hunk 3]
```diff
@@ -552,6 +579,23 @@ public static function validateProfileForm($values) {
...
+        if (
+          ! ctype_digit(strval($fieldValue)) 
+          || (int)$fieldValue > static::$_form_field_weight_max 
+        ) {
+          $errors[$fieldName] = E::ts(
+            'Weight field must be set to a integer number between 0 and %1 but has value [%2].',
+            array(
+              1 => static::$_form_field_weight_max,
+              2 => $fieldValue
+            )
```
Comment (jensschuppe) [inline CRM/Newsletter/Form/Profile.php:595]: Short array syntax, please.

File: CRM/Newsletter/Form/Profile.php [hunk 4]
```diff
@@ -552,6 +579,23 @@ public static function validateProfileForm($values) {
...
+          );
+        }
```
Comment (jensschuppe) [inline CRM/Newsletter/Form/Profile.php:597]: Why have a max and min weight at all …?

File: l10n/de_DE/LC_MESSAGES/newsletter.po [hunk 5]
```diff
@@ -236,6 +244,10 @@ msgstr "Mindestens ein Kontaktfeld muss aktiviert werden."
 msgid "Each active field needs a label."
 msgstr "Jedes aktive Feld benötigt eine Beschriftung."
 
+#: CRM/Newsletter/Form/Profile.php
+msgid "Each weight field must be set to a integer number between 0 and %1 but has value [%2]"
+msgstr "Für jede Position ist nur einen Wert zwischen 0 und %1 gültig, gesetzt wurde [%2]"
+
```
Comment (jensschuppe) [inline l10n/de_DE/LC_MESSAGES/newsletter.po:250]: The phrase should end with a period `.`

Comment (jensschuppe) [reply l10n/de_DE/LC_MESSAGES/newsletter.po:250]: Also, `s/einen/ein` in the German translation

Comment (jensschuppe) [reply l10n/de_DE/LC_MESSAGES/newsletter.po:250]: Also, `s/to a integer/to an integer`

File: templates/CRM/Newsletter/Form/Profile.hlp [hunk 6]
```diff
@@ -1,3 +1,17 @@
 {htxt id='unsubscribe_submit_label'}
   <p>{ts domain="de.systopia.newsletter"}This label is used if the profile contains only one list and the frontend is configured to hide the list name in that case.{/ts}</p>
 {/htxt}
+
+{assign var=max value=CRM_Newsletter_Form_Profile::$_form_field_weight_max} 
+{assign var=default value=CRM_Newsletter_Form_Profile::$_form_field_weight_default} 
+
+{htxt id='field_weight'}
+  <p>{ts domain="de.systopia.newsletter"}A weight is used to change the positioning of fields within the form.{/ts}</p>
+  <p>{ts domain="de.systopia.newsletter"}The effect of field positioning can be seen in the drupal view of this form.{/ts}</p>
```
Comment (jensschuppe) [inline templates/CRM/Newsletter/Form/Profile.hlp:10]: The *Drupal* implementation should not be mentioned here, as the API is supposed to work for any implementation, not just the Drupal module.

File: templates/CRM/Newsletter/Form/Profile.hlp [hunk 7]
```diff
@@ -1,3 +1,17 @@
...
+  <p>{ts domain="de.systopia.newsletter"}Leave weights empty for default positioning, that is, the position of the fields seen here.{/ts}</p>
+  <p>{ts domain="de.systopia.newsletter"}The default weight is always {$default}.{/ts}</p>
+  <p>{ts domain="de.systopia.newsletter"}If position of fields should be changed, choose values between 0 and {$max} for the fields in question.{/ts}</p>
+  <p>{ts domain="de.systopia.newsletter"}The higher the weight value, the more the field will be shifted to the top of the form.{/ts}</p>
+  <p>{ts domain="de.systopia.newsletter"}The lower the weight value, the more the field will be shifted to the bottom of the form.{/ts}</p>
```
Comment (jensschuppe) [inline templates/CRM/Newsletter/Form/Profile.hlp:15]: Actually, it's the other way around: higher weights cause fields to appear later in the form. Think things with *weight* sinking further down …

File: CRM/Newsletter/Form/Profile.php [hunk 8]
```diff
@@ -44,6 +44,20 @@ class CRM_Newsletter_Form_Profile extends CRM_Core_Form {
    * static::getXCMProfiles().
    */
   protected static $_xcm_profiles = NULL;
+  
+  /**
+   * @var string
+   *
+   * The default weight of form fields, for positioning within the form
+   */
+  public static $_form_field_weight_default = 50;
```
Comment (jensschuppe) [inline CRM/Newsletter/Form/Profile.php:53]: This is not being enforced, right? Why have a default weight at all? Let's just have no weight by default.

File: CRM/Newsletter/Form/Profile.php (hunk 3 above)
Comment (jensschuppe) [reply CRM/Newsletter/Form/Profile.php:595]: Also, let's apply Code standards - there should be no space after negating `!`, but after a type cast `(int)` …

**Review by jensschuppe — CHANGES_REQUESTED** [review]

**Comment by svenow** [comment]
New commits should have resolved all previous review remarks.

- the implicit default weight for all fields is 0
- no explicit min or max weight values are used
- as smaller/more negative a weight becomes, as higher the corresponding field moves up to the top of the form
- as larger/more positive a weight becomes, as lower the corresponding field moves down to the bottom of the form

File: CRM/Newsletter/Form/Profile.php [hunk 9]
```diff
@@ -552,6 +562,20 @@ public static function validateProfileForm($values) {
       ) {
         $errors['contact_field_' . $available_name . '_label'] = E::ts('Each active field needs a label.');
       }
+
+      $fieldName = 'contact_field_' . $available_name . '_weight';
+      $fieldValue = $values[$fieldName];
+      if (isset($fieldValue) 
+          && strlen($fieldValue) > 0 
+          && filter_var($fieldValue, FILTER_VALIDATE_INT) === false
```
Comment (jensschuppe) [inline CRM/Newsletter/Form/Profile.php:570]: `s/false/FALSE` according to CiviCRM Code standards.

**Review by jensschuppe — CHANGES_REQUESTED** [review]
Just a code review, did not test.

**Review by jensschuppe — APPROVED** [review]

**Pull request merged**